src/gflownet/tasks/seh_frag.py

```diff
@@ -0,0 +1,217 @@
+            'clip_grad_type': 'norm',
+            'clip_grad_param': 10,
+            'random_action_prob': 0.,
+            'sampling_tau': 0.,
+        }
+
+    def setup(self):
+        hps = self.hps
+        RDLogger.DisableLog('rdApp.*')
+        self.rng = np.random.default_rng(142857)
+        self.env = GraphBuildingEnv()
+        self.ctx = FragMolBuildingEnvContext(max_frags=9, num_cond_dim=32)
+        self.training_data = []
+        self.test_data = []
+        self.offline_ratio = 0
+        self.valid_offline_ratio = 0
+
+        model = GraphTransformerFragGFN(self.ctx, num_emb=hps['num_emb'], num_layers=hps['num_layers'])
+        self.model = model
+        # Separate Z parameters from non-Z to allow for LR decay on the former
+        Z_params = list(model.logZ.parameters())
+        non_Z_params = [i for i in self.model.parameters() if all(id(i) != id(j) for j in Z_params)]
+        self.opt = torch.optim.Adam(non_Z_params, hps['learning_rate'], (hps['momentum'], 0.999),
+                                    weight_decay=hps['weight_decay'], eps=hps['adam_eps'])
+        self.opt_Z = torch.optim.Adam(Z_params, hps['learning_rate'], (0.9, 0.999))
+        self.lr_sched = torch.optim.lr_scheduler.LambdaLR(self.opt, lambda steps: 2**(-steps / hps['lr_decay']))
+        self.lr_sched_Z = torch.optim.lr_scheduler.LambdaLR(self.opt_Z, lambda steps: 2**(-steps / hps['Z_lr_decay']))
+
+        self.sampling_tau = hps['sampling_tau']
+        if self.sampling_tau > 0:
+            self.sampling_model = copy.deepcopy(model)
+        else:
+            self.sampling_model = self.model
+        eps = hps['tb_epsilon']
+        hps['tb_epsilon'] = ast.literal_eval(eps) if isinstance(eps, str) else eps
+        self.algo = TrajectoryBalance(self.env, self.ctx, self.rng, hps, max_nodes=9)
+
+        self.task = SEHTask(self.training_data, hps['temperature_sample_dist'],
+                            ast.literal_eval(hps['temperature_dist_params']), wrap_model=self._wrap_model_mp)
+        self.mb_size = hps['global_batch_size']
+        self.clip_grad_param = hps['clip_grad_param']
+        self.clip_grad_callback = {
+            'value': (lambda params: torch.nn.utils.clip_grad_value_(params, self.clip_grad_param)),
+            'norm': (lambda params: torch.nn.utils.clip_grad_norm_(params, self.clip_grad_param)),
+            'none': (lambda x: None)
+        }[hps['clip_grad_type']]
+
+    def step(self, loss: Tensor):
+        loss.backward()
+        for i in self.model.parameters():
+            self.clip_grad_callback(i)
+        self.opt.step()
+        self.opt.zero_grad()
+        self.opt_Z.step()
+        self.opt_Z.zero_grad()
+        self.lr_sched.step()
+        self.lr_sched_Z.step()
+        if self.sampling_tau > 0:
+            for a, b in zip(self.model.parameters(), self.sampling_model.parameters()):
+                b.data.mul_(self.sampling_tau).add_(a.data * (1 - self.sampling_tau))
+
+
+# Determined-specific code:
```

Comment:
Should this live in gflownetR?

Reply:
Yes, missed that.

Reply:
Addressed by recursionpharma/gflownetR#14.